README.rst
=========
 Importer
=========

The Importer provides a way to call a remote procedure and retrieve its results
in a native format. It has to be used in conjunction with webengine/exporter.


License
=======

Importer is released under the `GNU LGPL 2.1 <http://www.gnu.org/licenses/lgpl-2.1.html>`_.


Build and installation
=======================

Bootstrapping
-------------

Importer uses the autotools for its build system.

If you checked out code from the git repository, you will need
autoconf and automake to generate the configure script and Makefiles.

To generate them, simply run::

    $ autoreconf -fvi

Building
--------

Importer builds like your typical autotools-based project::

    $ ./configure && make && make install


Development
===========

We use `semantic versioning <http://semver.org/>`_ for
versioning. When working on a development release, we append ``~dev``
to the current version to distinguish released versions from
development ones. This has the advantage of working well with Debian's
version scheme, where ``~`` is considered smaller than everything (so
version 1.10.0 is more up to date than 1.10.0~dev).


Authors
=======

Importer was started at SmartJog by Thomas Meson in 2009. Various employees and
interns from SmartJog fixed bugs and added features since then.

* Bryann Lamour
* Clément Bœsch
* Gilles Dartiguelongue
* Nicolas Noirbent
* Nicolas Delvaux <nicolas.delvaux@cji.paris>
* Philippe Bridant
* Rémi Cardona
* Thomas Meson


ChangeLog
Version 0.7.6 Released on 2023/01/10

	* Fix Json deserialization with newer Django versions
	* A bit of code modernization
	* Update copyright info

Version 0.7.5 Released on 2022/09/09

	* Workaround compatibility with Webengine running on Squeeze...

Version 0.7.4 Released on 2022/04/08

	* Fix existing plugins compatibility during msgpack deserialization

Version 0.7.3 Released on 2022/03/30

	* Fix Python3 deserialization with newer versions of msgpack

Version 0.7.2 Released on 2021/07/20

	* Format the code with Black

Version 0.7.1 Released on 2020/08/31

	* Fix Python3 deserialization format for msgpack
	* Format the code with Black

Version 0.7.0 Released on 2020/06/29

	* Re-release to fix upgrades

Version 0.6.6 Released on 2020/04/08

	* Port importer to python3

Version 0.6.5 Released on 2020/04/07

	* Modernize packaging

Version 0.6.4 Released on 2016/06/07

	* Fix hanging forever when remote server is unreachable

Version 0.6.3 Released on 2016/02/23

	* Improve compatibility with some existing exporter plugins

Version 0.6.2 Released on 2016/02/01

	* Fix a syntax error of a variable, headers should not contains underscores

Version 0.6.1 Released on 2015/12/15

	* Read raw data from responses
	  This restores the previous behaviour and fix a regression when data is
	  compressed.

Version 0.6.0 Released on 2015/12/07

	* Port from urllib2 to requests
	* Correct the style of the __init__.py.in
	* Remove a function not used
	* Change the default serializer method from pickle to msgpack

Version 0.5.0 Released on 2012/06/22

	* Set a default User-Agent and allow changing its value through
	  configuration.
	* Merge headers definitions in urllib2.Request call.
	* Mutualize body decompression code.
	* Handle missing trailing / from distant_url.
	* Add missing documentation.

Version 0.4.0 Released on 2012/05/28

	* Make importer support multiple serialization format.
	* Pylint/PEP8 pass on bin/importer.

Version 0.3.5 Released on 2011/04/22

	* Now released under the LGPL-2.1.

Version 0.3.4 Released on 2011/03/09

	* Actually do Content-Encoding checking in error handling,
	  closes: #3898.
	* Do not try to handle non-exporter-origined exceptions, closes: #4036.
	* Drop python 2.4 specific workarounds.

Version 0.3.3 Released on 2011/03/01

	* Decompress error message when http errors occur, closes: #3898.

Version 0.3.2 Released on 2011/01/19

	* Add bin/importer script.

Version 0.3.1 Released on 2010/12/14

	* Add missing NAME section to the man page.

Version 0.3.0 Released on 2010/12/09

	* Add gzip Accept-Encoding / Content-Encoding support, closes: #3717.

Version 0.2.6 Released on 2010/04/26

	* Add importer doc.

Version 0.2.5 Released on 2010/01/12

	* Set TCP timeout to 30 sec, add property to change it, closes: #2595.

Version 0.2.4 Released on 2009/12/09

	* Fix issues when using several Importer instances, closes: #2652.

Version 0.2.3 Released on 2009/10/27

	* Python 2.5 compatibility fix wrt urllib2 mem/fd leaks issues.

version 0.2.2 Released on 2009/06/26

	* Fix fd & memory leak on exception, closes: #2062.

Version 0.2.1 Released on 2009/06/23

	* Call only once install_opener.
	* Fix circular dependencies, closes: #2022.

Version 0.2.0 Released on 2009/06/22

	* Handle SSL certificates. Closes: #1912.
	* Fix: Display read message instead of UnpicklingError.
	* Allow calls to webengine methods.

Version 0.1.0 Released on 2009/05/12

	* Initial release
	* Added more exceptions class
	* Better error handling for UnpicklingError
	* Moved exceptions to their own python file
	* Fix bug: Different instances of Importer was sharing the same "conf"
	dictionary (due to the way python handle default arg for parameters)


importer.py
# -*- coding: utf-8 -*-

""" Importer """

import traceback, cookielib, urllib2, httplib, socket
import cStringIO
import gzip
from exc import *

__all__ = ['ImporterError', 'ImporterDeserializeError', 'ImporterSerializeError', 'ImporterConnectError', 'Importer']

class ImporterHTTPSHandler(urllib2.HTTPSHandler):
    """ To be used by urllib2, wrap calls to httplib. """
    def __init__(self, key, cert):
        urllib2.HTTPSHandler.__init__(self)
        self.key = key
        self.cert = cert

    def https_open(self, req):
        return self.do_open(self.create_conn, req)

    def create_conn(self, host, timeout = 300):
        return httplib.HTTPSConnection(host, key_file = self.key, cert_file = self.cert)

class ImporterBase(object):
    """
        Base class for both ImporterModule and ImporterVariable.
        Provide call/get/set methods, relying on proper self.__objinst__ object.
        ImporterModule and ImporterVariable are in charge of creating the appropriate object.
    """
    def __init__(self):
        self.__conf__ = {}
        self.__objinst__ = None

    def __getitem__(self, key):
        """
            Used to retrieve data from Importer configuration.
            Let exceptions be thrown.
        """
        return self.__conf__[key]

    def __setitem__(self, key, value):
        """ Set configuration key. """
        self.__conf__[key] = value

    def __delitem__(self, key):
        """ Delete configuration key. """
        del self.__conf__[key]

    def call(self, method, *args, **kw):
        """ Getattr self.__objinst__ which will be a module instance or a variable instance. """
        return getattr(self.__objinst__, method)(*args, **kw)

    def get(self, attr):
        """ Getattr self.__objinst__ and returns. """
        return getattr(self.__objinst__, attr)

class ImporterVariable(ImporterBase):
    def __init__(self, conf, module, klass, *args, **kw):
        ImporterBase.__init__(self)
        self.__mod__ = module
        self.__klass__ = klass
        self.__objinst__ = self.__mod__.get(self.__klass__)(*args, **kw)

class Importer(ImporterBase):
    """
        Main class. Contains all modules.
        Call/get/set/instantiate always check if execution must be done remotely.
    """
    def __init__(self):
        ImporterBase.__init__(self)
        self.__scope__ = {}
        self.__bound__ = None
        self.__cj__ = None

    def call(self, module, method, *args, **kw):
        """
            Perform a call to module.method, passing the given *args, **kw.
            Override ImporterBase.call()
            Return: module.method return.
        """
        if 'distant_url' in self.__conf__.keys():
            mod = '.'.join((module, method))
            return self.__perform_distant__(mod, 'call', *args, **kw)
        try:
            if module in self.__scope__.keys(): # Class instance
                return self.__scope__[module].call(method, *args, **kw)
            mod = __import__(module, {}, {}, [''])
            callee = getattr(mod, method)
            # The __request__ argument is added by the Exporter.
            # It is used only when the __exportable__ flag is present.
            request = kw.pop('__request__', None)
            # This attribute means that the callee object is a _Export object
            # defined by the @exportable decorator.
            # This is used to distinct between a standard python method, or a
            # webengine method, because a webengine method need a request object.
            if hasattr(callee, '__exportable__'): return callee(request, *args, **kw)
            else: return getattr(mod, method)(*args, **kw)
        except Exception, e:
            raise ImporterError(str(e), traceback=traceback.format_exc())

    def get(self, module, attr):
        """ Retrieve an attr from the given module. """
        if 'distant_url' in self.__conf__.keys():
            mod = '.'.join((module, attr))
            return self.__perform_distant__(mod, 'get')
        try:
            if module in self.__scope__.keys(): # Class instance
                return self.__scope__[module].get(attr)
            mod = __import__(module, {}, {}, [''])
            return getattr(mod, attr)
        except Exception, e:
            raise ImporterError(str(e), traceback=traceback.format_exc())

    def instantiate(self, variable, module, klass, *args, **kw):
        """
            Add in the current scope a 'klass' instance from module.
            Similar to module.klass(*args, **kw).
            variable will be usable in the scope as others modules.
        """
        if 'distant_url' in self.__conf__.keys():
            mod = '.'.join((module, klass))
            return self.__perform_distant__(mod, 'instantiate', variable=variable, *args, **kw)
        try:
            if variable in self.__scope__.keys(): return
            module = self.__load_module__(module)
            self.__scope__[variable] = ImporterVariable(self.__conf__, module, klass, *args, **kw)
        except Exception, e:
            raise ImporterError(str(e), traceback=traceback.format_exc())

    def _set_bound(self, bound):
        """ Bound Importer scope to "bound" list. """
        self.__bound__ = bound
    def _get_bound(self):
        """ Returns bound list. """
        return self.__bound__
    bound = property(_get_bound, _set_bound)

    def __perform_distant__(self, module, type, *args, **kw):
        """ Perform the distant call. """
        import cPickle
        try:
            if not self.__cj__:
                self.__cj__ = cookielib.CookieJar()
            if 'timeout' in self.__conf__.keys():
                socket.setdefaulttimeout(self.__conf__['timeout'])
            else:
                socket.setdefaulttimeout(30)
            opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(self.__cj__), ImporterHTTPSHandler(key = self.__conf__.get('ssl_key'), cert = self.__conf__.get('ssl_cert')))
            path = module.replace('.', '/') + '/' #Force trailing slash
            # Should be able to select encoder
            # TODO: Create a wrapper for cPickle, pickle in fallback
            data = cPickle.dumps({'type': type, 'args': args, 'kw': kw}, cPickle.HIGHEST_PROTOCOL)
            req = urllib2.Request(self.__conf__['distant_url'] + path, data, {'WEBENGINE_OUTPUT': 'pickle'})
            # We accept gzipped content
            req.add_header('Accept-Encoding', 'gzip')
            f = opener.open(req)
            if f.headers.get('Content-Encoding') == 'gzip':
                # Decompress the data
                data_compressed = cStringIO.StringIO(f.read())
                gzip_decompressor = gzip.GzipFile(fileobj = data_compressed)
                data_read = gzip_decompressor.read()
            else:
                data_read = f.read()
            if data_read == '': return None
            try:
                data_decoded = cPickle.loads(data_read)
                return data_decoded
            except cPickle.UnpicklingError:
                raise ImporterDeserializeError(data_read, traceback=traceback.format_exc())
        except urllib2.HTTPError, e:
            if e.headers.get('Content-Encoding') == 'gzip':
                # Decompress the data
                data_compressed = cStringIO.StringIO(e.read())
                gzip_decompressor = gzip.GzipFile(fileobj = data_compressed)
                error = gzip_decompressor.read()
            else:
                error = e.read()

            # Most likely a normal HTTP server error
            if e.headers.get('Content-Type') != 'application/octet-stream':
                raise ImporterError(str(e), local=False)

            try:
                data_decoded = cPickle.loads(error) # Read exception
            except cPickle.UnpicklingError, e:
                raise ImporterDeserializeError(error, traceback=error)
            raise ImporterError(data_decoded['msg'], local=False, traceback=data_decoded['traceback'])
        except urllib2.URLError, e:
            raise ImporterConnectError(str(e), traceback=traceback.format_exc())
        except cPickle.PickleError, e:
            raise ImporterSerializeError(str(e), traceback=traceback.format_exc())


importer/decoders.py
# Decoders


class Decoder:

    mime_type = None

    def __init__(self):
        self._obj = None
        self._input = "None"

    def decode(self, obj):
        self._obj = obj

    def __str__(self):
        if self._obj:
            return self.decode(self._obj)
        return "<Decoder %s>" % str(self._input)


class DefaultDecoder(Decoder):

    mime_type = "text/plain"

    def __init__(self):
        super().__init__()
        self._input = "default"

    def decode(self, obj):
        super().decode(obj)
        return str(obj)


class XmlDecoder(Decoder):

    mime_type = "text/xml"

    def __init__(self):
        super().__init__()
        self._input = "xml"

    def decode(self, obj):
        super().decode(obj)
        # TODO


class SoapDecoder(Decoder):

    mime_type = "application/soap+xml"

    def __init__(self):
        super().__init__()
        self._input = "soap"

    def decode(self, obj):
        super().decode(obj)
        # TODO


class JSONDecoder(Decoder):

    mime_type = "application/json"

    def __init__(self):
        super().__init__()
        self._input = "json"

    def decode(self, obj):
        super().decode(obj)
        import json

        return json.JSONDecoder().decode(obj.decode())


class PickleDecoder(Decoder):

    mime_type = "application/pickle"

    def __init__(self):
        super().__init__()
        self._input = "pickle"

    def decode(self, obj):
        super().decode(obj)
        import pickle

        return pickle.loads(obj)


class MsgpackDecoder(Decoder):

    mime_type = "application/x-msgpack"

    def __init__(self):
        super().__init__()
        self._input = "msgpack"

    def decode(self, obj):
        super().decode(obj)
        import msgpack

        return msgpack.unpackb(obj, raw=False, strict_map_key=False)


class DecoderFactory:
    """DecoderFactory handles Decoder instances for the right input mode."""

    _input = {
        "default": DefaultDecoder,
        "xml": XmlDecoder,
        "soap": SoapDecoder,
        "json": JSONDecoder,
        "pickle": PickleDecoder,
        "msgpack": MsgpackDecoder,
    }

    @staticmethod
    def get(in_decoder):
        """
        Returns a valid Decoder instance for the given input mode.
        If the input mode is invalid, returns a DefaultDecoder.
        """

        decoder = "default"
        if in_decoder in DecoderFactory._input:
            decoder = in_decoder
        return DecoderFactory._input[decoder]()


importer/generators.py
# Generator


class Generator:

    mime_type = None

    def __init__(self):
        self._obj = None
        self._output = "None"

    def generate(self, obj):
        self._obj = obj

    def __str__(self):
        if self._obj:
            return self.generate(self._obj)
        return "<Generator %s>" % str(self._output)


class XmlGenerator(Generator):

    mime_type = "text/xml"

    def __init__(self):
        super().__init__()
        self._output = "xml"

    def generate(self, obj):
        super().generate(obj)
        # TODO


class SoapGenerator(Generator):

    mime_type = "application/soap+xml"

    def __init__(self):
        super().__init__()
        self._output = "soap"

    def generate(self, obj):
        super().generate(obj)
        # TODO


class JSONGenerator(Generator):

    mime_type = "application/json"

    def __init__(self):
        super().__init__()
        self._output = "json"

    def generate(self, obj):
        super().generate(obj)
        import json
        from django.core.serializers.json import DjangoJSONEncoder

        return json.dumps(obj, cls=DjangoJSONEncoder)


class PickleGenerator(Generator):

    mime_type = "application/pickle"

    def __init__(self):
        super().__init__()
        self._output = "pickle"

    def generate(self, obj):
        super().generate(obj)
        import pickle

        try:
            return pickle.dumps(obj, pickle.HIGHEST_PROTOCOL)
        except pickle.PickleError as _error:
            # cPickle failed, try pickle
            import pickle

            return pickle.dumps(obj, pickle.HIGHEST_PROTOCOL)


class MsgpackGenerator(Generator):

    mime_type = "application/x-msgpack"

    def __init__(self):
        super().__init__()
        self._output = "msgpack"

    def generate(self, obj):
        super().generate(obj)
        import msgpack

        # use_bin_type is required while we need to exchange with very old versions in production.
        # This should be removable once everyone uses Python3.
        return msgpack.packb(obj, use_bin_type=False, default=str)


class DefaultGenerator(Generator):

    mime_type = "text/plain"

    def __init__(self):
        super().__init__()
        self._output = "default"

    def generate(self, obj):
        super().generate(obj)
        return str(obj)


class GeneratorFactory:
    """GeneratorFactory handles Generator instances for the right output mode."""

    _output = {
        "xml": XmlGenerator,
        "soap": SoapGenerator,
        "json": JSONGenerator,
        "pickle": PickleGenerator,
        "default": DefaultGenerator,
        "msgpack": MsgpackGenerator,
    }

    @staticmethod
    def get(output):
        """
        Returns a valid Generator instance for the given output mode.
        If the output mode is invalid, returns a DefaultGenerator.
        """

        out = "default"
        if output in GeneratorFactory._output:
            out = output
        return GeneratorFactory._output[out]()


importer/__init__.py
# -*- coding: utf-8 -*-

""" Importer """

import traceback, cookielib, urllib2, httplib, socket
import cStringIO
import gzip
from importer.exc import ImporterError, ImporterDeserializeError, ImporterSerializeError, ImporterConnectError
from importer.decoders import DecoderFactory
from importer.generators import GeneratorFactory

__all__ = ['ImporterError', 'ImporterDeserializeError', 'ImporterSerializeError', 'ImporterConnectError', 'Importer']

class ImporterHTTPSHandler(urllib2.HTTPSHandler):
    """ To be used by urllib2, wrap calls to httplib. """
    def __init__(self, key, cert):
        urllib2.HTTPSHandler.__init__(self)
        self.key = key
        self.cert = cert

    def https_open(self, req):
        return self.do_open(self.create_conn, req)

    def create_conn(self, host, timeout = 300):
        return httplib.HTTPSConnection(host, key_file = self.key, cert_file = self.cert)

class ImporterBase(object):
    """
        Base class for both ImporterModule and ImporterVariable.
        Provide call/get/set methods, relying on proper self.__objinst__ object.
        ImporterModule and ImporterVariable are in charge of creating the appropriate object.
    """
    def __init__(self):
        self.__conf__ = {}
        self.__objinst__ = None

    def __getitem__(self, key):
        """
            Used to retrieve data from Importer configuration.
            Let exceptions be thrown.
        """
        return self.__conf__[key]

    def __setitem__(self, key, value):
        """ Set configuration key. """
        self.__conf__[key] = value

    def __delitem__(self, key):
        """ Delete configuration key. """
        del self.__conf__[key]

    def call(self, method, *args, **kw):
        """ Getattr self.__objinst__ which will be a module instance or a variable instance. """
        return getattr(self.__objinst__, method)(*args, **kw)

    def get(self, attr):
        """ Getattr self.__objinst__ and returns. """
        return getattr(self.__objinst__, attr)

class ImporterVariable(ImporterBase):
    def __init__(self, conf, module, klass, *args, **kw):
        ImporterBase.__init__(self)
        self.__mod__ = module
        self.__klass__ = klass
        self.__objinst__ = self.__mod__.get(self.__klass__)(*args, **kw)

class Importer(ImporterBase):
    """
        Main class. Contains all modules.
        Call/get/set/instantiate always check if execution must be done remotely.
    """
    def __init__(self):
        ImporterBase.__init__(self)
        self.__scope__ = {}
        self.__bound__ = None
        self.__cj__ = None

    def call(self, module, method, *args, **kw):
        """
            Perform a call to module.method, passing the given *args, **kw.
            Override ImporterBase.call()
            Return: module.method return.
        """
        if 'distant_url' in self.__conf__.keys():
            mod = '.'.join((module, method))
            return self.__perform_distant__(mod, 'call', *args, **kw)
        try:
            if module in self.__scope__.keys(): # Class instance
                return self.__scope__[module].call(method, *args, **kw)
            mod = __import__(module, {}, {}, [''])
            callee = getattr(mod, method)
            # The __request__ argument is added by the Exporter.
            # It is used only when the __exportable__ flag is present.
            request = kw.pop('__request__', None)
            # This attribute means that the callee object is a _Export object
            # defined by the @exportable decorator.
            # This is used to distinct between a standard python method, or a
            # webengine method, because a webengine method need a request object.
            if hasattr(callee, '__exportable__'): return callee(request, *args, **kw)
            else: return getattr(mod, method)(*args, **kw)
        except Exception, e:
            raise ImporterError(str(e), traceback=traceback.format_exc())

    def get(self, module, attr):
        """ Retrieve an attr from the given module. """
        if 'distant_url' in self.__conf__.keys():
            mod = '.'.join((module, attr))
            return self.__perform_distant__(mod, 'get')
        try:
            if module in self.__scope__.keys(): # Class instance
                return self.__scope__[module].get(attr)
            mod = __import__(module, {}, {}, [''])
            return getattr(mod, attr)
        except Exception, e:
            raise ImporterError(str(e), traceback=traceback.format_exc())

    def instantiate(self, variable, module, klass, *args, **kw):
        """
            Add in the current scope a 'klass' instance from module.
            Similar to module.klass(*args, **kw).
            variable will be usable in the scope as others modules.
        """
        if 'distant_url' in self.__conf__.keys():
            mod = '.'.join((module, klass))
            return self.__perform_distant__(mod, 'instantiate', variable=variable, *args, **kw)
        try:
            if variable in self.__scope__.keys(): return
            module = self.__load_module__(module)
            self.__scope__[variable] = ImporterVariable(self.__conf__, module, klass, *args, **kw)
        except Exception, e:
            raise ImporterError(str(e), traceback=traceback.format_exc())

    def _set_bound(self, bound):
        """ Bound Importer scope to "bound" list. """
        self.__bound__ = bound
    def _get_bound(self):
        """ Returns bound list. """
        return self.__bound__
    bound = property(_get_bound, _set_bound)

    def __perform_distant__(self, module, type, *args, **kw):
        """ Perform the distant call. """
        try:
            if not self.__cj__:
                self.__cj__ = cookielib.CookieJar()
            if 'timeout' in self.__conf__.keys():
                socket.setdefaulttimeout(self.__conf__['timeout'])
            else:
                socket.setdefaulttimeout(30)
            opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(self.__cj__), ImporterHTTPSHandler(key = self.__conf__.get('ssl_key'), cert = self.__conf__.get('ssl_cert')))
            path = module.replace('.', '/') + '/' #Force trailing slash
            # Serialize with serialization method provided in configuration
            serializer = GeneratorFactory.get(self.__conf__.get('serialization', 'pickle'))
            data = serializer.generate({'type': type, 'args': args, 'kw': kw})
            req = urllib2.Request(self.__conf__['distant_url'] + path, data, {'WEBENGINE_OUTPUT': self.__conf__.get('serialization', 'pickle')})
            # We accept gzipped content
            req.add_header('Accept-Encoding', 'gzip')
            req.add_header('Content-Type', serializer.mime_type)
            f = opener.open(req)
            if f.headers.get('Content-Encoding') == 'gzip':
                # Decompress the data
                data_compressed = cStringIO.StringIO(f.read())
                gzip_decompressor = gzip.GzipFile(fileobj = data_compressed)
                data_read = gzip_decompressor.read()
            else:
                data_read = f.read()
            if data_read == '': return None
            try:
                serializer = DecoderFactory.get(self.__conf__.get('serialization', 'pickle'))
                data_decoded = serializer.decode(data_read)
                return data_decoded
            except Exception:
                raise ImporterDeserializeError(data_read, traceback=traceback.format_exc())
        except urllib2.HTTPError, e:
            if e.headers.get('Content-Encoding') == 'gzip':
                # Decompress the data
                data_compressed = cStringIO.StringIO(e.read())
                gzip_decompressor = gzip.GzipFile(fileobj = data_compressed)
                error = gzip_decompressor.read()
            else:
                error = e.read()

            # Most likely a normal HTTP server error
            if e.headers.get('Content-Type') != 'application/octet-stream':
                raise ImporterError(str(e), local=False)

            try:
                serializer = DecoderFactory.get(self.__conf__.get('serialization', 'pickle'))
                data_decoded = serializer.decode(error)  # Read exception
            except Exception, e:
                raise ImporterDeserializeError(error, traceback=error)
            raise ImporterError(data_decoded['msg'], local=False, traceback=data_decoded['traceback'])
        except urllib2.URLError, e:
            raise ImporterConnectError(str(e), traceback=traceback.format_exc())
        except Exception, e:
            raise ImporterSerializeError(str(e), traceback=traceback.format_exc())
